qutip/tests/core/test_qobjevo.py

```diff
@@ -42,111 +42,442 @@
+    tlistlog = np.logspace(-3, 1, 10001)
+    args = {'w1': 1, "w2": 2}
+    cte_qobj = rand_herm(N)
+    real_qobj = Qobj(np.random.rand(N, N))
+    cplx_qobj = rand_herm(N)
+    real_coeff = Base_coeff(_real, "sin(t*w1)")
+    cplx_coeff = Base_coeff(_cplx, "exp(1j*t*w2)")
+    qobjevos = {"qobj": rand_herm(N), "scalar":np.random.rand()}
+    cplx_forms = ['func', 'string', 'spline', 'array', 'array_log']
+    mixed_forms = ['with_args', 'no_args', 'mixed_tlist']
+
+    def _rand_t(self):
+        return np.random.rand() * 0.9 + 0.05
+
+    def test_creation(self, form):
+        if form == "cte":
+            obj = QobjEvo(self.cte_qobj)
+        elif form == "func":
+            obj = QobjEvo([self.cte_qobj,
+                           [self.cplx_qobj, self.cplx_coeff.func]],
+                          args=self.args)
+        elif form == "string":
+            obj = QobjEvo([self.cte_qobj,
+                           [self.cplx_qobj, self.cplx_coeff.string]],
+                          args=self.args)
+        elif form == "spline":
+            obj = QobjEvo([[self.cplx_qobj,
+                            self.cplx_coeff.spline(self.tlist, self.args)],
+                           self.cte_qobj])
+        elif form == "array":
+            obj = QobjEvo([self.cte_qobj,
+                           [self.cplx_qobj,
+                            self.cplx_coeff.array(self.tlist, self.args)]],
+                          tlist=self.tlist)
+        elif form == "array_log":
+            obj = QobjEvo([self.cte_qobj,
+                           [self.cplx_qobj,
+                            self.cplx_coeff.array(self.tlistlog, self.args)]],
+                          tlist=self.tlistlog)
+        elif form == "with_args":
+            obj = QobjEvo([self.cte_qobj,
+                           [self.cplx_qobj, self.cplx_coeff.func],
+                           [self.real_qobj, self.real_coeff.string]],
+                          args=self.args)
+        elif form == "no_args":
+            obj = QobjEvo([self.cte_qobj,
+                           [self.cplx_qobj, self.cplx_coeff.spline(self.tlist,
+                                                                   self.args)],
+                           [self.real_qobj, self.real_coeff.array(self.tlist,
+                                                                  self.args)]],
+                          tlist=self.tlist)
+        elif form == "mixed_tlist":
+            coeff_1 = coefficient(self.cplx_coeff.array(self.tlist,
+                                                        self.args),
+                                  tlist=self.tlist)
+            coeff_2 = coefficient(self.real_coeff.array(self.tlistlog,
+                                                        self.args),
+                                  tlist=self.tlistlog)
+            obj = QobjEvo([self.cte_qobj,
+                           [self.cplx_qobj, coeff_1],
+                           [self.real_qobj, coeff_2]])
```

Comment:
This `if .. elif .. elif` isn't really how the parametrisation is meant to be used.  You could parametrise over a tuple of `form, input_args, input_kwargs` parameters or something (you can override the default fixture using `mark.parametrize`).  The advantage of this is that `pytest` will print out the values of the input parameters on a test failure, so we'd see them right there in the test log, whereas here we'd just get the string, and it takes more work to find out what's gone on.

Second, there's no assertions in this test; I get that it's in theory just testing that there are no errors on construction, but it should probably also assert that the output has the correct number of ops, the coefficients are what we expect, etc etc.